# How many ways of disconnenting the London Tube exist

The London Tube can be naturally seen as a connected graph (this is so natural that is the [official representation](https://en.wikipedia.org/wiki/Tube_map). This connectiveness is a property that is used on a daily basis for every commuter in London. 

But **how weak is this transport network?** We say that a node in a graph is an articulation vertex if its removal [disconnects the graph](https://mathworld.wolfram.com/ArticulationVertex.html). We are going to count how many articulation vertices does the London Tube have.

For this little experiment we are going to use the help of the libraries Pandas and NetworkX.

In [10]:
import pandas as pd
import networkx as nx

Thanks to [this library](https://github.com/nicola/tubemaps) we have CSVs for the london tube lines, stations and connections. We place them inside three Pandas DataFrames:

In [2]:
lines       = pd.read_csv('london.lines.csv', index_col=0)
stations    = pd.read_csv('london.stations.csv', index_col=0)
connections = pd.read_csv('london.connections.csv')

In [3]:
lines.head(3)

name  colour stripe
line                                        
1               Bakerloo Line  AE6017    NaN
3                 Circle Line  FFE02B    NaN
6     Hammersmith & City Line  F491A8    NaN

In [4]:
stations.head(3)

latitude  longitude          name       display_name  zone  total_lines  \
id                                                                            
1    51.5028    -0.2801    Acton Town    Acton<br />Town   3.0            2   
2    51.5143    -0.0755       Aldgate                NaN   1.0            2   
3    51.5154    -0.0726  Aldgate East  Aldgate<br />East   1.0            2   

    rail  
id        
1      0  
2      0  
3      0

In [5]:
connections.head(3)

station1  station2  line  time
0        11       163     1     1
1        11       212     1     2
2        49        87     1     1

After that we build up a graph:

In [13]:
graph = nx.Graph()

for connection_id, connection in connections.iterrows():
    station1_name = stations.loc[connection['station1']]['name']
    station2_name = stations.loc[connection['station2']]['name']
    graph.add_edge(station1_name, station2_name, time = connection['time'])

From this representation we can see how in fact the graph is connected

In [14]:
nx.is_connected(graph)

True

It is now the time to ask ourselves what are the articulation vertices of the London Tube?

In [15]:
list(nx.articulation_points(graph))

['Kenton',
 'South Kenton',
 'North Wembley',
 'Wembley Central',
 'Stonebridge Park',
 'Harlesden',
 'Willesden Junction',
 'Kensal Green',
 "Queen's Park",
 'Kilburn Park',
 'Maida Vale',
 'Warwick Avenue',
 'Paddington',
 'Ruislip Gardens',
 'South Ruislip',
 'Northolt',
 'Greenford',
 'Perivale',
 'Hanger Lane',
 'North Acton',
 'Kew Gardens',
 'Gunnersbury',
 'Turnham Green',
 'Theydon Bois',
 'Debden',
 'Loughton',
 'Buckhurst Hill',
 'Woodford',
 'Leytonstone',
 'Leyton',
 'Stratford',
 'Gallions Reach',
 'Cyprus',
 'Beckton Park',
 'Royal Albert',
 'Prince Regent',
 'Custom House',
 'Royal Victoria',
 'Canning Town',
 'Elverson Road',
 'Deptford Bridge',
 'Greenwich',
 'Cutty Sark',
 'Island Gardens',
 'Mudchute',
 'Crossharbour & London Arena',
 'South Quay',
 'Heron Quays',
 'Canary Wharf',
 'Shadwell',
 'Surrey Quays',
 'Canada Water',
 'South Wimbledon',
 'Colliers Wood',
 'Tooting Broadway',
 'Tooting Bec',
 'Balham',
 'Clapham South',
 'Clapham Common',
 'Clapham North',


How many are they ?

In [16]:
len(list(nx.articulation_points(graph)))

142

Hence there exist $2^{142} = 5575186299632655785383929568162090376495104$ ways of disrupting the network. Quite a lot aren't they?